### PR TITLE
Support for Zipkin Protobuf spans over HTTP

### PR DESCRIPTION
**Pull request by jan25**
## Which problem is this PR solving?
- Adds support in collector for Zipkin Protobuf spans sent to `/api/v2/spans`
- Fixes https://github.com/jaegertracing/jaeger/issues/1586

## Short description of the changes
- We had JSON support for Zipkin spans sent over HTTP to v2 endpoint, this PR is a mirrored implementation based on [jsonv2](https://github.com/jaegertracing/jaeger/blob/master/cmd/collector/app/zipkin/jsonv2.go) implementation
- Related PRs that should be merged to complete adding zipkin proto support: https://github.com/jaegertracing/xdock-zipkin-brave/pull/13 and https://github.com/jaegertracing/jaeger-idl/pull/56
- <s>Temporarily marking this PR as WIP. Please review/suggest on the implementation</s>

### CODE REVIEW

File: cmd/collector/app/zipkin/protov2.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+		}
+		tSpans = append(tSpans, tSpan)
+	}
+	return tSpans, nil
+}
+
+func protoSpanV2ToThrift(s *model.SpanModel) (*zipkincore.Span, error) {
+	sc := s.SpanContext
+	ts := nanoToMicroSecs(s.Timestamp.UnixNano())
+	d := nanoToMicroSecs(s.Duration.Nanoseconds())
```
Comment (yurishkuro) [inline cmd/collector/app/zipkin/protov2.go]: possible loss of precision

**Comment by jan25** [comment]
@yurishkuro Thanks!
Logic-wise, no difference except the type definitions changing a little. I thought zipkin-go is officially supported/made for different collector like Jaeger.

You have a fair point about us having to control which proto version we want to support. I can update this PR to use [proto](https://github.com/openzipkin/zipkin-api/blob/master/zipkin.proto) file directly and generate code, and [jaeger-idl](https://github.com/jaegertracing/jaeger-idl) looks like ideal place to put this proto file.

**Comment by yurishkuro** [comment]
Yes, the proto file can go to jaeger-idl

**Comment by pavolloffay** [comment]
> Manual test the v2 endpoint with zipkin proto spans

Have a look at xdock tests, there are already tests for zipkin thrift and v2 json. We could leverage the same approach and add proto. It will require adding support for proto reporter in https://github.com/jaegertracing/xdock-zipkin-brave

**Comment by jan25** [comment]
Looks like CI is stuck as below, `make fmt` on my computer didn't update any source files. Any hint is appreciated.

```Running go fmt on ALL_SRC ...
Go fmt, license check, or import ordering failures, run 'make fmt'
./cmd/collector/app/zipkin/http_handler_test.go
make: *** [lint] Error 1
The command "if [ "$TESTS" == true ]; then make test-ci ; else echo 'skipping tests'; fi" exited with 2.```

**Comment by yurishkuro** [comment]
try to run `make install-tools` first, then `make fmt`. I am getting a change in cmd/collector/app/zipkin/http_handler_test.go.

Also, make sure you're on Go 1.12, as linting behavior changes between versions.

I pushed the formatting change to your branch.

**Comment by jan25** [comment]
I missed `make install-tools`, thanks for the fix!

File: cmd/collector/app/zipkin/protov2.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+			return nil, err
+		}
+		tSpans = append(tSpans, tSpan)
+	}
+	return tSpans, nil
+}
+
+func protoSpanV2ToThrift(s *zipkinProto.Span) (*zipkincore.Span, error) {
+	if len(s.Id) != model.TraceIDShortBytesLen {
+		return nil, fmt.Errorf("invalid Span ID")
```
Comment (pavolloffay) [inline cmd/collector/app/zipkin/protov2.go]: nit: make it more specific? Invalid length of the span ID?

**Comment by jan25** [comment]
@pavolloffay thanks! Added them now. I think we should merge https://github.com/jaegertracing/xdock-zipkin-brave/pull/13 to go forward with this main PR.

**Comment by pavolloffay** [comment]
@jan25 I have added there one comment, otherwise it looks good to me.

File: cmd/collector/app/zipkin/protov2.go [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+		tSpan.BinaryAnnotations = append(tSpan.BinaryAnnotations, &zipkincore.BinaryAnnotation{
+			Key:            zipkincore.LOCAL_COMPONENT,
+			Host:           localE,
+			AnnotationType: zipkincore.AnnotationType_STRING,
+		})
+	}
+	return tSpan, nil
+}
+
+func traceIDFromBytes(tid []byte) (model.TraceID, error) {
```
Comment (yurishkuro) [inline cmd/collector/app/zipkin/protov2.go]: Perhaps incorporate this into `func (t *TraceID) Unmarshal(data []byte) error` instead. In which case the `*BytesLen` constants can remain private

**Comment by pavolloffay** [comment]
thanks @jan25

**Comment by jan25** [comment]
Thanks for all the help! @yurishkuro I was working on your comment in the review :) but this is already merged, will open different PR when i have that refactored.

**Comment by pavolloffay** [comment]
I have considered it as fixed due to :+1: mark on the comment. thanks @jan25

**Comment by pavolloffay** [comment]
@jan25 are you using Jaeger in booking.com? If yes could you please add an entry to https://github.com/jaegertracing/jaeger/blob/master/ADOPTERS.md or at least comment in https://github.com/jaegertracing/jaeger/issues/207

**Comment by jan25** [comment]
@pavolloffay Not yet, But noted the links from your previous comment to update them when booking.com does :) thanks!

**Comment by pavolloffay** [comment]
thanks @jan25 !

**Pull request merged**